File: src/main/java/core/procedures/cons/Cdr.java
package core.procedures.cons;

import core.procedures.AFn;
import core.scm.FnArgs;
import core.scm.ICons;
import core.scm.SCMClass;

import java.util.List;

@FnArgs(minArgs = 1, maxArgs = 1, mandatoryArgsTypes = {SCMClass.SCMPair.class})
public class Cdr extends AFn {

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "cdr";
  }

  @Override
  public Object apply1(Object arg) {
    return cdr(arg);
  }

  public static Object cdr(Object o) {
    if (o instanceof ICons) {
      return ((ICons)o).cdr();
    }
    List list = (List) o;
    return list.subList(1, list.size());
  }
}

File: src/main/kotlin/core/procedures/cons/Cdr.kt
package core.procedures.cons

import core.procedures.AFn
import core.scm.Cons
import core.scm.Type

class Cdr : AFn<Any?, Any>(name = "cdr", isPure = true, minArgs = 1, maxArgs = 1,
                            mandatoryArgsTypes = arrayOf<Class<*>>(Type.Pair::class.java)) {

    override operator fun invoke(arg: Any?) = when (arg) {
        is Cons<*> -> arg.cdr()
        else       -> (arg as List<*>).subList(1, arg.size)
    }
}